File: src/engine/adapters/s3/transfer.ts
import { randomUUID } from 'node:crypto';
import { createReadStream, createWriteStream } from 'node:fs';
import { rename, stat, unlink } from 'node:fs/promises';
import type { Readable } from 'node:stream';
import { pipeline } from 'node:stream/promises';
import {
  GetObjectCommand,
  type GetObjectCommandOutput,
  HeadObjectCommand,
  type S3Client,
} from '@aws-sdk/client-s3';
import type { ObjectTransferResult } from '@shared/domain/object-store';
import { OBJECT_UPLOAD_MAX_BYTES } from '@shared/protocol/page';
import type { OpCtx } from '../adapter';
import { AdapterError, throwIfCancelled } from '../errors';
import { mapError } from './errors';
import { formatBytes } from './read';

// P33: the only file in this adapter that touches node:fs — Adapter rule 1 ("an adapter imports
// nothing from electron") allows a plain Node module free rein here, and F15 confirms the engine
// utilityProcess can do its own file I/O without routing bytes through main.

// D10: writes to a sibling temp file and renames on success, unlinking on any failure or
// cancellation — streaming straight to destPath would leave a truncated file wearing the name of
// a complete one if the transfer is cancelled or fails partway through, with no way for the user
// to tell. Same directory as destPath, so the rename is atomic on one filesystem.
export async function downloadObject(
  client: S3Client,
  bucket: string,
  key: string,
  destPath: string,
  ctx: OpCtx,
): Promise<ObjectTransferResult> {
  throwIfCancelled(ctx);
  ctx.setCommand(`GetObject s3://${bucket}/${key} -> ${destPath}`);

  // HeadObject first — a real "no such object" error before any local file is even created,
  // mirroring read.ts's own HeadObject-before-GetObject discipline.
  try {
    await client.send(new HeadObjectCommand({ Bucket: bucket, Key: key }), {
      abortSignal: ctx.signal,
    });
  } catch (err) {
    throw mapError(err);
  }

  const tmpPath = `${destPath}.kira-partial-${randomUUID()}`;
  let res: GetObjectCommandOutput;
  try {
    res = await client.send(new GetObjectCommand({ Bucket: bucket, Key: key }), {
      abortSignal: ctx.signal,
    });
  } catch (err) {
    throw mapError(err);
  }
  if (!res.Body) throw new AdapterError('E_QUERY', `${key} has no body to download`);

  try {
    await pipeline(res.Body as Readable, createWriteStream(tmpPath), { signal: ctx.signal });
    const written = await stat(tmpPath);
    await rename(tmpPath, destPath);
    return { bytes: written.size };
  } catch (err) {
    await unlink(tmpPath).catch(() => {});
    throw mapError(err);
  }
}

// stat()s the file before any network call, so a missing/unreadable source or an over-limit file
// fails before PutObject ever starts. AWS's own single-PutObject ceiling (D12) — above it an
// upload needs multipart, which this phase does not implement; the message names the limit rather
// than failing silently or truncating.
export async function openUploadBody(
  sourcePath: string,
): Promise<{ stream: Readable; size: number }> {
  let size: number;
  try {
    const info = await stat(sourcePath);
    size = info.size;
  } catch (err) {
    throw new AdapterError(
      'E_QUERY',
      `could not read local file ${sourcePath}: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
  if (size > OBJECT_UPLOAD_MAX_BYTES) {
    throw new AdapterError(
      'E_UNSUPPORTED',
      `file is ${formatBytes(size)}, over the ${formatBytes(OBJECT_UPLOAD_MAX_BYTES)} single-upload limit — multipart upload is not supported`,
    );
  }
  return { stream: createReadStream(sourcePath), size };
}

File: shell/internal/adapters/s3/transfer.go
package s3

import (
	"context"
	"io"
	"os"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/google/uuid"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
	"github.com/kirathecat/kira-studio/shell/internal/page"
	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// downloadObject is transfer.ts's downloadObject — the phase's first real file transfer, and the
// only code in P58 that writes to a user's filesystem. Every step is transfer.ts translated, in
// the same order, for the same reasons (P58d §4.4):
//
//  1. CheckNotStarted: an already-cancelled op never creates a file.
//  2. op.SetCommand before any call (Adapter rule 3).
//  3. HeadObject first — a real "no such object" error before any local file exists.
//  4. tmp is a SIBLING of destPath, never os.TempDir() — the rename must be atomic on one
//     filesystem.
//  5. Create the temp file.
//  6. GetObject.
//  7. io.Copy streams the body in — ctx cancellation surfaces here, mid-stream, since the op's own
//     ctx goes directly to the SDK call (P58d D3: never adapters.RunWithAbortRace, which would let
//     a cancelled download keep writing after the caller's cleanup already ran).
//  8. Sync and close.
//  9. Rename — the last step, never retried; a cross-device rename is impossible by construction
//     (step 4's sibling-temp-file rule), so there is no fallback copy on purpose.
//
// On any error from steps 5-9, the temp file is unlinked (error ignored) before the mapped error
// is returned — and that cleanup runs on the caller's own goroutine, which is only safe because
// the copy itself never outlives this call (P58d D3 again).
func downloadObject(ctx context.Context, client *s3.Client, bucket, key, destPath string, op *adapters.OpCtx) (model.ObjectTransferResult, error) {
	if err := adapters.CheckNotStarted(ctx); err != nil {
		return model.ObjectTransferResult{}, err
	}
	op.SetCommand("GetObject s3://" + bucket + "/" + key + " -> " + destPath)

	if _, err := client.HeadObject(ctx, &s3.HeadObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)}); err != nil {
		return model.ObjectTransferResult{}, mapError(err)
	}

	tmpPath := destPath + ".kira-partial-" + uuid.NewString()
	f, err := os.Create(tmpPath)
	if err != nil {
		return model.ObjectTransferResult{}, mapError(err)
	}
	cleanup := func() { _ = os.Remove(tmpPath) }

	res, err := client.GetObject(ctx, &s3.GetObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
	if err != nil {
		f.Close()
		cleanup()
		return model.ObjectTransferResult{}, mapError(err)
	}
	if res.Body == nil {
		f.Close()
		cleanup()
		return model.ObjectTransferResult{}, mapError(errNoBody(key))
	}
	defer res.Body.Close()

	n, err := io.Copy(f, res.Body)
	if err != nil {
		f.Close()
		cleanup()
		return model.ObjectTransferResult{}, mapError(err)
	}
	if err := f.Sync(); err != nil {
		f.Close()
		cleanup()
		return model.ObjectTransferResult{}, mapError(err)
	}
	if err := f.Close(); err != nil {
		cleanup()
		return model.ObjectTransferResult{}, mapError(err)
	}
	if err := os.Rename(tmpPath, destPath); err != nil {
		cleanup()
		return model.ObjectTransferResult{}, mapError(err)
	}

	return model.ObjectTransferResult{Bytes: n}, nil
}

type queryError string

func (e queryError) Error() string { return string(e) }

func errNoBody(key string) error { return queryError(key + " has no body to download") }

// openUploadBody is transfer.ts's openUploadBody — stat()s the file before any network call, so a
// missing/unreadable source or an over-limit file fails before PutObject ever starts. The file is
// seekable and its length known (P58d D6), which lets the SDK checksum by seeking rather than
// falling into the aws-chunked path a non-seekable body would need. Closed by the caller with a
// defer (mutate.go's applyInsert) — the TypeScript never closes its createReadStream because the
// SDK consumes it and Node's GC finishes the job, but a leaked fd in Go is real.
func openUploadBody(sourcePath string) (*os.File, int64, error) {
	info, err := os.Stat(sourcePath)
	if err != nil {
		return nil, 0, adapters.New(adapters.CodeQuery, "could not read local file "+sourcePath+": "+err.Error(), err)
	}
	if info.Size() > int64(page.ObjectUploadMaxBytes) {
		return nil, 0, adapters.New(adapters.CodeUnsupported,
			"file is "+formatBytes(info.Size())+", over the "+formatBytes(int64(page.ObjectUploadMaxBytes))+" single-upload limit — multipart upload is not supported", nil)
	}
	f, err := os.Open(sourcePath)
	if err != nil {
		return nil, 0, adapters.New(adapters.CodeQuery, "could not read local file "+sourcePath+": "+err.Error(), err)
	}
	return f, info.Size(), nil
}
